doxygen/contributor_help_pages/common_pitfalls.md

```diff
@@ -0,0 +1,239 @@
+  Eigen::VectorXd b = Eigen::VectorXd(a.size());
+  return make_holder([](const auto& l, const auto& r) { return l + r; }, std::move(b));
+}
+```
+
+Returning expressions is an advanced feature and it is easy to make
+mistakes. In this regard, it is simplest to start development not
+returning expressions (in this case holders are unnecessary), and only
+add expression return types later.
+
+It is always possible to return a non-expression type by evaluating
+the Eigen expression. For convenience, there is an `eval` function
+in Math that will evaluate Eigen expressions and forward anything
+else. This is convenient because it works on non-Eigen types as well
+(as compared to the built in `.eval()` member function on Eigen
+types).
+
+The implementation of `make_holder` is
+[here](https://github.com/stan-dev/math/blob/develop/stan/math/prim/meta/holder.hpp).
+
+### Move Semantics
+
+In general, Stan math does not use move semantics very often. This is because of our arena allocator. Move semantics generally work as
+
+```cpp
+
+class my_big_type {
+  double* m_data;
+
+  explicit my_big_type(size_t N) m_data(static_cast<double*>(malloc(N))) {}
+  // Standard copy constructor
+  my_big_type(const my_big_type& other) m_data(static_cast<double*>(malloc(N))) {
+    std::copy(other.m_data, this->m_data);
+  }
+  // Standard move constructor
+  my_big_type(my_big_type&& other) m_data(other.m_data) {
+    other.m_data = nullptr;
+  }
+};
+```
+
+We can see in the above that the standard style of a move (the constructor taking an rvalue reference) is to copy the pointer and then null out the original pointer. But in Stan, particularly for reverse mode, we need to keep memory around even if  it's only a temporary for when we call the gradient calculations in the reverse pass. And since memory for reverse mode is stored in our arena allocator no copying happens in the first place.
+
+When working with arithmetic types, keep in mind that moving Scalars is often less optimal than simply taking their copy. For instance, Stan's `var` type is a PIMPL implimentation, so it simply holds a pointer of size 8 bytes. A `double` is also 8 bytes which just so happens to fit exactly in a [word](https://en.wikipedia.org/wiki/Word_(computer_architecture)) of most modern CPUs. While a reference to a double is also 8 bytes, unless the function is inlined by the compiler, the computer will have to place the reference into cache, then go fetch the value that is being referenced which now takes up two words instead of one!
+
+The general rules to follow for passing values to a function are:
+
+1. If a type is smaller than two words (16 bytes), pass it by value
+2. If you are writing a function for reverse mode, pass values by `const&`
+3. In prim, if you are confident and working with larger types, use perfect forwarding to pass values that can be moved from. Otherwise simply pass values by `const&`.
+
+### Copying non-arena variables to lambdas used in the reverse pass
```

Comment:
@bbbales2 can you add make_callback_ptr() stuff here?

Reply:
Done!